tools/astarte_export/lib/export.ex

```diff
@@ -0,0 +1,348 @@
+         {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"astarte", []}, state),
+         {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"devices", []}, state),
+         {:ok, conn} <- FetchData.db_connection_identifier(),
+         {:ok, state} <- process_devices(conn, realm, fd, state),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state),
+         {:ok, _state} <- XMLGenerate.xml_write_end_tag(fd, state),
+         :ok <- File.close(fd) do
+      Logger.info("Export Completed.", realm: realm, tag: "export_completed")
+      {:ok, :export_completed}
+    else
+      {:error, reason} -> 
+        File.close(fd) 
+        {:error, reason}
+    end
+  end
+
+  defp process_devices(conn, realm, fd, state) do
+    tables_page_configs = Application.get_env(:xandra, :cassandra_table_page_sizes, [])
+    page_size = Keyword.get(tables_page_configs, :device_table_page_size, 100)
+   options = [page_size: page_size]
+    process_devices(conn, realm, fd, state, options)
+  end
+
+  defp process_devices(conn, realm, fd, state, options) do
+    with {:more_data, device_list, updated_options} <-
+           FetchData.fetch_device_data(conn, realm, options),
+         {:ok, state} <- process_device_list(conn, realm, device_list, fd, state),
+         {:ok, paging_state} when paging_state != nil <-
+           Keyword.fetch(updated_options, :paging_state) do
+      process_devices(conn, realm, fd, state, updated_options)
+    else
+      {:ok, nil} -> {:ok, state}
+      {:ok, completed} -> {:ok, state}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+
+
+  defp process_device_list(_, _, [], _, state) do
+    {:ok, state}
+  end
+
+  defp process_device_list(conn, realm, [h | t] = device_list, fd, state) do
+    with {:ok, state} <- do_process_device(conn, realm, h, fd, state) do
+      process_device_list(conn, realm, t, fd, state)
+    end
+  end
+
+
+
+  defp do_process_device(conn, realm, device_data, fd, state) do
+    mapped_device_data = FetchData.process_device_data(device_data)
+    device = mapped_device_data.device
+    with {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"device", device}, state),
+         {:ok, state} <- construct_device_xml_tags(mapped_device_data, fd, state),
+         {:ok, state} <- process_interfaces(conn, realm, device_data, fd, state),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      {:ok, state}
+    end
+  end
+
+  def process_interfaces(conn, realm, device_data, fd, state) do
+    with {:ok, interfaces} <- FetchData.get_interface_details(conn, realm, device_data),
+         {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"interfaces", []}, state),
+         {:ok, state} <- process_interface_list(conn, realm, interfaces, fd, state),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      {:ok, state} 
+    end
+  end
+
+
+
+  defp process_interface_list(_, _, [], _, state) do
```

Comment:
By the way, this is not state, it is an accumulator

Reply:
Because, Accumulator would be more generic term.

As we are maintaining the state of the xml_tags history, it would be better to continue with state as attribute name